Relay Client Auth & other crypto specs (#605)
WalletConnect/walletconnect-docs: docs/specs/clients/core/crypto/crypto-authentication.md

# Crypto Authentication

In this module we will handle also authentication payloads used by multiple high-level SDKs which require an Ed25519 key pair to sign did-jwt payloads for multiple purposes.

## DID-JWT

> [DID-JWT](https://github.com/decentralized-identity/did-jwt) are [JSON Web Tokens (JWT)](https://jwt.io/) which are self-issued using cryptographic key pairs in elliptic curves such as `secp256k` or `ed25519`.

Reviewer: Maybe we only reference ED25519 as to not confuse anyone into thinking we also use secp256k.

```suggestion
[DID-JWT](https://github.com/decentralized-identity/did-jwt) are [JSON Web Tokens (JWT)](https://jwt.io/) which are self-issued using cryptographic key pairs in elliptic curves such as `ed25519`.
```

As per the JWT spec, all fields are serialized as base64url encoding and separated by full stops (`.`).

```sh
serialized_did_jwt = "eyJhbGciOiJFZERTQSIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJkaWQ6a2V5Ono2TWtvZEhad25lVlJTaHRhTGY4SktZa3hwREdwMXZHWm5wR21kQnBYOE0yZXh4SCIsInN1YiI6ImM0NzlmZTVkYzQ2NGU3NzFlNzhiMTkzZDIzOWE2NWI1OGQyNzhjYWQxYzM0YmZiMGI1NzE2ZTViYjUxNDkyOGUiLCJhdWQiOiJ3c3M6Ly9yZWxheS53YWxsZXRjb25uZWN0LmNvbSIsImlhdCI6MTY1NjkxMDA5NywiZXhwIjoxNjU2OTk2NDk3fQ.bAKl1swvwqqV_FgwvD4Bx3Yp987B9gTpZctyBviA-EkAuWc8iI8SyokOjkv9GJESgid4U8Tf2foCgrQp2qrxBA"
```

### Header

> The header will allow you to identify the JWT type and the algorithm used which in our protocols we use exclusively `ed25519` which is labelled as `EdDSA` in the header.

Reviewer: ```suggestion
The header will allow you to identify the JWT type and the algorithm used. We exclusively  use `ed25519` which is labelled as `EdDSA` in the header.
```

```jsonc
// header
{
  "alg": "EdDSA",
  "typ": "JWT"
}
```

> Public keys are resolved using the Decentralized ID (DID) of the signing identity of the token, which is passed as the iss attribute of the JWT payload. We will exclusively identify them to as [did:key](https://w3c-ccg.github.io/did-method-key/)

Reviewer 2: ```suggestion
Public keys are resolved using the Decentralized ID (DID) of the signing identity of the token, which is passed as the `iss` attribute of the JWT payload. We will exclusively identify them to as [did:key](https://w3c-ccg.github.io/did-method-key/)
```

> Public keys are resolved using the Decentralized ID (DID) of the signing identity of the token, which is passed as the iss attribute of the JWT payload. We will be usually exclusively [did:key](https://w3c-ccg.github.io/did-method-key/) to identify them

Reviewer: ```suggestion
Public keys are resolved using the Decentralized ID (DID) of the signing identity of the token, which is passed as the iss attribute of the JWT payload. We will usually be exclusive to identify them as [did:key](https://w3c-ccg.github.io/did-method-key/)
```

```sh
public_key = "did:key:z6MkiTBz1ymuepAQ4HEHYSF1H8quG5GLVVQR3djdX3mDooWp"
```

### Payload

> The payload will include multiple claims and the context can be identifiable by identifying the first claim `act` which will describe the payload's action/intent.

Reviewer: ```suggestion
The payload will include multiple claims. The context can be identified by the first claim `act`, which will describe the payload's action/intent.
```

Our protocols will share the following claims with different context:

`act` - description of action intent

`iat` - timestamp when jwt was issued

`exp` - timestamp when jwt must expire

`iss` - did:key of the key used to sign

`aud` - it varies (usually the recipient of the payload)

`sub` - it varies (usually provides context about the payload)

### Signature

> The header and payload must each be stringified JSON and then concatenated as bytes with the header first followed by the payload this will then be signed with the private key.

Reviewer: ```suggestion
The header and payload must each be stringified JSON and then concatenated as bytes with the header first followed by the payload. This will be signed with the private key.
```
